pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+	})
+
+	for _, cn := range newCandidates {
+		// If the NodePool isn't using WhenEmpty, checking if it's an invalid candidate can't be done quickly.
+		if nodePools[cn.Labels()[v1beta1.NodePoolLabelKey]].Spec.Disruption.ConsolidationPolicy != v1beta1.ConsolidationPolicyWhenEmpty {
+			continue
+		}
+		pods, err := cn.ReschedulablePods(ctx, c.kubeClient)
+		if err != nil {
+			return fmt.Errorf("getting candidate pods, %w", err)
+		}
+		if len(pods) > 0 {
+			return fmt.Errorf("found non-empty node with NodePool consolidationPolicy set to %q", v1beta1.ConsolidationPolicyWhenEmpty)
+		}
+	}
+
+	// Cross reference each node to its nodepool:
+	// 1. check if consolidation is enabled
+	// 2. check if node disruption budgets allowed disruptions is 0
+	for npName := range nodePoolsToCandidates {
+		nodePool := nodePools[npName]
+		if nodePool.Spec.Disruption.ConsolidateAfter.Duration == nil {
+			return fmt.Errorf("consolidation is not enabled for nodepool %s", npName)
+		}
+		// If the nodepool has no allowed disruptions, we know we can't disrupt this node
+		if c.cluster.AllowedNodePoolDisruptions(ctx, npName, c.clk) == 0 {
+			return fmt.Errorf("nodepool %q has no allowed disruptions", npName)
+		}
+	}
+
+	return nil
+}
+
+// computeConsolidation optimistically tries to find a consolidation decision:
+//  1. Iterate through the candidate nodes one at a time
+//     a. If the node cannot be consolidated, continue to the next node in the list.
+//  2. If the currently considered node can be consolidated, start increasing the batch size.
+//  3. Increase the batch size by a factor of 2 for every valid scheduling simulation.
+//     a. With a max batch size of 100, this means the max number of scheduling simulations for each starting point will be
+//     1 -> 2 -> 4 -> 8 -> 16 -> 32 -> 64 -> 100 = 8
+func (c *Controller) computeConsolidation(ctx context.Context) ([]*state.StateNode, error) {
+	candidates := c.Candidates(ctx)
+	lastValidMax := -1
+	start := c.clk.Now()
+	for i := range candidates {
+		// add a timeout for finding a single valid command.
+		if c.clk.Since(start) > singleTimeout {
+			ConsolidationTimeoutTotalCounter.Inc()
+			break
+		}
+		batchSize := 1
+		multiBatchStart := c.clk.Now()
+		// Get the subset of candidates at indices [i, i + batchSize), and do a scheduling simulation. If we succeed, double
+		// the size of the subset by doubling the batch size, where the batch size will be <= len(candidates) and <= 100.
+		for ; batchSize <= maxBatchSize; batchSize = lo.Clamp(batchSize*2, 1, maxBatchSize) {
+			// add a timeout on increasing the batch size.
+			if c.clk.Since(multiBatchStart) > multiTimeout {
+				ConsolidationTimeoutTotalCounter.Inc()
+				break
+			}
+			maxIndex := lo.Clamp(i+batchSize, 0, len(candidates))
+			nodesToConsolidate := lo.Slice(candidates, i, maxIndex)
+			results, err := SimulateScheduling(ctx, c.kubeClient, c.cluster, c.provisioner, c.recorder, nodesToConsolidate...)
```

Comment:
What are the scenarios where we return errors from scheduling and do we always want to break out of consolidation for all of them?